### DISCUSSION TITLE
Error in AMSRE CDF generation

### DISCUSSION THREAD

**Discussion by nchawang** [NASA-LIS/LISF#998] [Q&A] [answered]
Hi LIS community,

I was following the steps from the test case for data assimilation of soil moisture (https://lis.gsfc.nasa.gov/lis-testcases), I wanted to assimilated NASA-AMSRE soil moisture for the period of 2010. Steps 3 and 4 i was able to complete without any issue, but in step 5, while trying to generate the CDF for AMSR-E i got the following error.

"Program received signal SIGFPE: Floating-point exception - erroneous arithmetic operation.

Backtrace for this error:
#0 0x3878c3299f in ???
#1 0x66cbcd in ???
#2 0x6953ec in ???
#3 0x5a11a8 in ???
#4 0x4a8971 in ???
#5 0x4a94ab in ???
#6 0x51ef65 in ???
#7 0x56a171 in ???
#8 0x594aa3 in ???
#9 0x594bec in ???
#10 0x3878c1ed1c in ???
#11 0x407aa4 in ???
Floating point exception (core dumped)"
I downloaded the AMSR-E files from https://nsidc.org/data/AE_Land3/versions/2

I'm not sure why this is happening, if anyone knows please help, I have attached the ldtlog.0000 and ldt.config. LDT 7.3 was compiled in serial with HDF4 and HDFEOS.  

[CDF_AMSRE_Error.zip](https://github.com/NASA-LIS/LISF/files/8152107/CDF_AMSRE_Error.zip)

**Comment by dmocko** [comment, marked as answer]
Hi,

Can you please attach your configure.ldt file?

The end of your ldtlog file reads:

```
 [INFO] Reading NCEP Native Monthly Albedo 
 #Regridding: Applying bilinear interp to input parameter
 Done reading /data/noel/lis_param/ldt_param/albedo/albedo_jul.asc
 Reading /data/noel/lis_param/ldt_param/albedo/albedo_aug.asc
```

Since LDT stopped here, it may indicate a problem with this file.  Is it the same size as the other albedo files?

David

**Reply by nchawang** [reply]
Hi David,
Thankyou for the response, the albedo files are the same size as the other, while performing the CDF for LIS OL run i had faced no issues with it. 
please find the configure.ldt : [configure.ldt.txt](https://github.com/NASA-LIS/LISF/files/8166859/configure.ldt.txt)
I have recompiled with LIS7.4

Additionally while trying to understand where the code is stopping, i had added some print statements (Attached).

- So first i added the print statements to NASA_AMSREsm_obsMod.F90 and recompiled LDT which showed the following output when i ran LDT with the same earlier config file.

file : [NASA_AMSREsm_obsMod.F90.txt](https://github.com/NASA-LIS/LISF/files/8166919/NASA_AMSREsm_obsMod.F90.txt)
output :
_"NASA AMSRE soil moisture directory found
NASA AMSRE GRID set to EASE
NASA AMSRE GRID limits set
NASA AMSRE GRID decription set 
NASA AMSRE GRID decription set ML 1
npts calculated
NASA AMSRE npts set
NASA AMSRE npts allocated
NASA AMSRE obs%n112 set to 0.0"_

then the same errors appear, It did not print the next print line which was "neighbor interpolate NASA AMSRE" so then that meant it stopped at function 
call neighbor_interp_input(n, NASA_AMSREsmobs%gridDesci, &
         NASA_AMSREsmobs%n112)

- To understand further, after this i added some print statements in ./interp/neighbor_interp_input.F90 and recompiled LDT and ran it again.

file : [neighbor_interp_input.F90.txt](https://github.com/NASA-LIS/LISF/files/8166921/neighbor_interp_input.F90.txt)
output : 
_"(print statements of neighbor_interp_input from other files)
(print statements from NASA_AMSREsm_obsMod.F90)
....
NASA AMSRE npts set
NASA AMSRE npts allocated
NASA AMSRE obs%n112 set to 0.0
staring call
finish call
loop 1 101022
in if condition 953 267 9.0000000 1383.0000000 586.0000000 -90.0000000 -179.609604 0.0000000 83.3378830 180.130096 1.0000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000
"_
then the error, it did not print the next statement of "after allocation" in the neighbot_interp_input.F90 file. This pointed to the error being somewhere at
        n112(n)=get_fieldpos(i1,j1,gridDesci)

- This then made me add some more print statements in the ./interp/get_fieldpos.F90 files and recompiled LDT. 

file : [get_fieldpos.F90.txt](https://github.com/NASA-LIS/LISF/files/8166978/get_fieldpos.F90.txt)
output :
_".... (print statements of neighbor_interp_input from other files)
....(print statements from NASA_AMSREsm_obsMod.F90 and neighbor_interp_input.F90)
finish call
loop 1 101022
in if condition 953 267 9.0000000 1383.0000000 586.0000000 -90.0000000 -179.609604 0.0000000 83.3378830 180.130096 1.0000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000
dlon = 0.00000000
variables are : -179.609604 180.130096 0 0 1383
im ig condition 1 met"_

- then the error comes again, I'm still not sure where inexact pin point is the error coming from but, here were few things i noticed, ig is 0 (calculated as ig=nint(360/abs(dlon)) ) probably because dlon is coming to be 0.0000000 this is because of the if condition get_fieldpos.F90.txt

_"  if(gridDesc(1).eq.0) then 
     dlon = gridDesc(9)
  elseif(gridDesc(1).eq.1) then 
     dlon = gridDesc(8)
  elseif(gridDesc(1).eq.4) then 
     dlon = gridDesc(9)
  elseif(gridDesc(1).eq.9) then 
     dlon = gridDesc(10)
  endif"_

gridDesc when traced back is defined in NASA_AMSREsm_obsMod.F90 (after removal of print statements) as 

"NASA_AMSREsmobs%gridDesci(1) = 9  !input is EASE grid
 !these  corner coordinates were calculated based on ezlh_convert
 NASA_AMSREsmobs%gridDesci(4) = -90.0  !lat
 NASA_AMSREsmobs%gridDesci(5) = -179.6096 !lon
 NASA_AMSREsmobs%gridDesci(7) = 83.33788  !lat
 NASA_AMSREsmobs%gridDesci(8) = 180.1301  !lon    
 NASA_AMSREsmobs%gridDesci(2) = ease_nc  !nx
 NASA_AMSREsmobs%gridDesci(3) = ease_nr  !ny    
 NASA_AMSREsmobs%gridDesci(9)  = 1  !Ml
 npts= LDT_rc%lnc(n)*LDT_rc%lnr(n)
 NASA_AMSREsmobs%mo=npts
 allocate(NASA_AMSREsmobs%n112(npts))
 NASA_AMSREsmobs%n112=0.0
 call neighbor_interp_input(n, NASA_AMSREsmobs%gridDesci, &
       NASA_AMSREsmobs%n112)"

because it falls under the if-else condition **if(gridDesc(1).eq.9)** it takes **dlon = gridDesc(10)** which is 0 as per the print statement (_print*, 'in if condition', i1,j1,gridDesci_) i had added in neighbot_interp_input.F90. I dont know if this plays a role, please help.

**Reply by nchawang** [reply]
Hi David,
So i have added two more print statement in ./interp/get_fieldpos.F90 in lines 88 onwards as
file : [get_fieldpos.F90.txt](https://github.com/NASA-LIS/LISF/files/8167358/get_fieldpos.F90.txt)

_"ig=nint(360/abs(dlon))
print*,'variables are : ',rlon1,rlon2,iscan,ig,im
if(im.ge.ig) then
**print*,'im ig condition 1 met', im, ig, i**
ii=mod(i-1+ig,ig)+1
**print*, 'ii =',ii**
if((j.le.0.or.j.ge.jm+1).and.mod(ig,2).eq.0) then"_

The output then comes as:
_"...(earlier prints)
finish call
loop 1 101022
in if condition 953 267 9.0000000 1383.0000000 586.0000000 -90.0000000 -179.609604 0.0000000 83.3378830 180.130096 1.0000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000
dlon = 0.00000000
variables are : -179.609604 180.130096 0 0 1383
im ig condition 1 met 1383 0 953"_

But not the next print statement of **print*, 'ii =',ii**, this would mean the problem is in 
**ii=mod(i-1+ig,ig)+1**
from the print statement, **im=1383, ig=0 and i=953**.
so that means the denominator is being fed in as 0 in the mod function.
ig is calculated as : **ig=nint(360/abs(dlon))**
and **dlon is also 0** as per the if-else condition in which since **gridDesc(1).eq.9**, **dlon = gridDesc(10)** which is 0.

gridDesc is defined in NASA_AMSREsm_obsMod.F90 as per the end of the earlier comment, I could not find 
**NASA_AMSREsmobs%gridDesci(10) being defined anywhere. Although NASA_AMSREsmobs%gridDesci(9) is defined as 1**  
was this supposed to be NASA_AMSREsmobs%gridDesci(10), what is NASA_AMSREsmobs%gridDesci(9) for?

Any help would be appreciated.

**Reply by dmocko** [reply]
Hi @nchawang 

Can you please let us know the name of the AMSRE file you are trying to read?

Based on the link that you provided, the filename convention is: AMSR_E_L3_DailyLand_V07_20111003.hdf

However, this line is trying to read the V06 version of these files:

https://github.com/NASA-LIS/LISF/blob/af94ed88100f392c1b066d2aa743e0e64d27af44/ldt/DAobs/NASA_AMSRE_sm/readNASA_AMSREsmObs.F90#L441

Can you change the code on this line from "V06" to "V07" and re-compile?

It looks like a new version of this data came out, and we were not aware of it yet.

David

**Reply by nchawang** [reply]
Hi @dmocko 
Apologies for the late reply
V07 was changed from 27-8-2010 from (https://nsidc.org/data/AE_Land3/versions/2). I ran the CDF generation for the year 2010, after renaming the files from V06 to V07, still the same problem persists.
I tried running for 6 months CDF gen for Jan 1, 2010 to Jun 30, 2010. The same problem persists. 
I also did the same for 1 month CDF gen from the RS data in the Data portal for LIS (https://portal.nccs.nasa.gov/lisdata_pub/data/RS_DATA/AMSRE_SM/) the same problem rises.

Thanks
Noel C.

**Reply by dmocko** [reply]
We are finding the same error, and are looking into a cause and a solution.

David

**Reply by nchawang** [reply]
Okay,
Thankyou

**Reply by dmocko** [reply]
Hi @nchawang 

Sorry for the delay.  Please try the following to correct this crash:

In LISF/ldt/DAobs/NASA_AMSRE_sm/NASA_AMSREsm_obsMod.F90, add the following line of code:
    `NASA_AMSREsmobs%gridDesci(10)  = 0.25`

You can add it near the lines where the other %gridDesci values are defined.

Re-compile and re-run, and it should work.

Many thanks to @yhkwon81 for investigating this and for providing a solution.

I am going to label this as a bug, and open up an Issue with a fix.

**Reply by nchawang** [reply]
Hi @dmocko,
Apologies for the late reply.
This worked, thankyou! (and @yhkwon81 )
I added it to ldt, recompiled and checked the cdf graph, everything is normal.

For the lis portion I added the same in LISF/lis/dataassim/obs/NASA_AMSREsm/NASA_AMSREsm_Mod.F90 and recompiled it.

When I ran LIS filling in details from the same https://lis.gsfc.nasa.gov/sites/default/files/LIS/public_7_3_releases/LIS_usersguide.31-mar-2021.html#ssec_dataassim following the test case after CDF generation.

I get the following error

"***********************Error************************************
setup routine for DA obs AMSR-E(NASA) soil moisture is not defined
program will seg fault.....
***********************Error************************************"

after checking, in https://lis.gsfc.nasa.gov/sites/default/files/LIS/public_7_3_releases/LIS_usersguide.31-mar-2021.html#_build_instructions section 5.4.1 "Customizing the build via LIS plugins" I do not find any DA OBS on NASA AMSRE sm in Table 11 to know if it is on/off by default, is that a plugin I have to activate separately?

**Reply by bmcandr** [reply]
Hi @nchawang,

Looks like the NASA AMSRE SM obs reader is missing from our `lis/make/default.cfg` so this plugin is not being compiled. Adding the following to the end of the DA obs section of `default.cfg` will allow our build system to pick up this plugin

```
[DA OBS NASA_AMSREsm]
enabled: True
macro: DA_OBS_NASA_AMSRESM
path: dataassim/obs/NASA_AMSREsm
```

**BUT**, compiling results in the following errors:

```
../dataassim/obs/NASA_AMSREsm/NASA_AMSREsm_Mod.F90(267): error #6404: This name does not have a type, and must have an explicit type.   [LIS_ENSONGRID]
               grid=LIS_ensOnGrid(n),name="Observation"//vid(1)//vid(2),&
--------------------^
../dataassim/obs/NASA_AMSREsm/NASA_AMSREsm_Mod.F90(266): error #6284: There is no matching specific function for this generic function reference.   [ESMF_FIELDCREATE]
          pertField(n) = ESMF_FieldCreate(arrayspec=pertArrSpec,&
-------------------------^
compilation aborted for ../dataassim/obs/NASA_AMSREsm/NASA_AMSREsm_Mod.F90 (code 1)
make: *** [NASA_AMSREsm_Mod.o] Error 1
```

As far as I can tell, the `LIS_ensOnGrid` routine mentioned in the error message no longer exists. This routine is called in a number of DA OBS readers that are listed in our docs, but also missing from `make/default.cfg` (e.g., ANSA SWE, SMOSL2sm). This seems to suggest they were disabled and may no longer be supported...

@dmocko, what do you think?

**Reply by bmcandr** [reply]
Hi again,

I've identified changes you can make that will allow the NASA AMSRE Soil Moisture DA obs plugin to compile:

1. Add the following to `make/default.cfg`:
    ```
    [DA OBS NASA_AMSREsm]
    enabled: True
    macro: DA_OBS_NASA_AMSRESM
    path: dataassim/obs/NASA_AMSREsm
    ```

2. Edit  `dataassim/obs/NASA_AMSREsm/NASA_AMSREsm_Mod.F90` to make the following changes:
    Add `use LIS_DAobservationsMod` on line 81
    `LIS_vecGrid(n)` --> `LIS_obsvecGrid(n,k)`
    `LIS_ensOnGrid(n)` --> `LIS_obsensongrid(n,k)`

Full disclosure: we do not have a testcase for this plugin so while LIS will now compile, I cannot guarantee this plugin will run or function properly. We have reached out to our DA gurus for their inputs and I will follow up with their thoughts.

**Reply by nchawang** [reply]
Hi @bmcandr 

Thankyou for the reply, I made the changes to the files make/default.cfg and dataassim/obs/NASA_AMSREsm/NASA_AMSREsm_Mod.F90 and compiled it. The compilation worked, but like you said, the plugin did not. it still shows the same error which is

"Error*************
setup routine for DA obs AMSR-E(NASA) soil moisture is not defined
program will seg fault.....
Error*************"

is the name nomenclature for NASA AMSRE soil moisture in the "Data assimilation set" of the lis.config file (https://lis.gsfc.nasa.gov/sites/default/files/LIS/public_7_3_releases/LIS_usersguide.31-mar-2021.html, section 9.4)  different to that of the User guide?

**Reply by dmocko** [reply]
Hi @nchawang 

Would you mind copying your lis.config file, so we can try to run it as well?

I believe the issue is in LISF/lis/plugins/LIS_DAobs_pluginMod.F90.

You will see these lines:

```
#if 0 
   use NASA_AMSREsm_Mod,        only : NASA_AMSREsm_setup
#endif
```

```
#if 0
   external read_NASA_AMSREsm, write_NASA_AMSREsmobs
#endif
```

```
#if 0
!NASA AMSRE obs 
   call registerdaobsclass(trim(LIS_NASA_AMSREsmobsId),"LSM")
   call registerdaobssetup(trim(LIS_NASA_AMSREsmobsId)//char(0),NASA_AMSREsm_setup)
   call registerreaddaobs(trim(LIS_NASA_AMSREsmobsId)//char(0),read_NASA_AMSREsm)
   call registerwritedaobs(trim(LIS_NASA_AMSREsmobsId)//char(0),write_NASA_AMSREsmobs)
#endif
```

Change the "#if 0" lines to:  "#if [DA OBS NASA_AMSREsm]".

Go into "LISF/lis/make", type "gmake realclean", and then re-configure/re-compile.

**Reply by nchawang** [reply]
Hi @dmocko 

So I changed the "#if 0" to "#if [DA OBS NASA_AMSREsm]"; it showed me the following error:
"#if [DA OBS NASA_AMSREsm]
Error : token "[" is not valid in preprocessor expressions
../plugins/LIS_DAobs_pluginMod.F90:334.0:

#if [DA OBS NASA_AMSREsm]
Error : token "[" is not valid in preprocessor expressions
../plugins/LIS_DAobs_pluginMod.F90:547.0:

#if [DA OBS NASA_AMSREsm]
Error : token "[" is not valid in preprocessor expressions
make: *** [LIS_DAobs_pluginMod.o] Error 1"

looking at the options above I changed "#if 0" to "#if ( define DA_OBS_NASA_AMSRE_SM )" it does not show the above error and compile is successful however, when I run LIS post compilation, the same error comes again, i.e.,

"Error*************
setup routine for DA obs AMSR-E(NASA) soil moisture is not defined
program will seg fault.....
Error*************"

PFA the lis.config file as well
[lis.config.txt](https://github.com/NASA-LIS/LISF/files/8407080/lis.config.txt)

**Reply by dmocko** [reply]
Hi @nchawang 

Sorry for giving you the incorrect syntax in my previous suggestion.  The correct syntax should actually be:
`#if ( defined DA_OBS_NASA_AMSRESM )`

Please correct it in the three instances in lis/plugins/LIS_DAobs_pluginMod.F90.

Be sure to go into "LISF/lis/make" and run "gmake realclean" again before re-compiling.

Again, we will note that the NASA_AMSRE soil moisture DA plugin is disabled.  Can you instead try the SMAP or the ESACCI soil moisture DA instances in LIS?

Also, can you please let us know if you have any collaborative projects with any members of the LIS team for this effort?

David

**Reply by nchawang** [reply]
Hi @dmocko 

Thanks I have made the change and it compiled successfully, however, when I run the LIS the following error comes
At line 5544 of file ../core/LIS_historyMod.F90 (unit = 14, file = './../../CDF_gen/LSM_CDF/cdf_noah36.nc')
Fortran runtime error: End of file

Error termination. Backtrace:
#0 0x7fc3d0aa4f7f in us_read
at ../.././libgfortran/io/transfer.c:2519
#1 0x7fc3d0aa5065 in pre_position
at ../.././libgfortran/io/transfer.c:2641
#2 0x7fc3d0aa77fc in data_transfer_init
at ../.././libgfortran/io/transfer.c:3135

I tried ESACCI, everything went well, the run was successful, with the new compiled LIS as well. however with NASA AMSRE the compile was successful, the CDF gen with LDT was successful (from the earlier given solution) but in run the above error shows up. I used the same LSM cdf as I did for the ESA CCI run.

I have worked on EaKF but it was mainly on atmospheric modelling, more recently I have begun looking at Land surface models and capabilities of the LIS with DA. I do not have any collaborative projects with any members of the LIS team, but am open to.
Thanks

**Reply by dmocko** [reply]
I would recommend that you use ESA CCI soil moisture data assimilation.  It's clear that the NASA AMSR-E code will take plenty of effort to modernize with the current code.  It seems your error may originate from where LIS is trying to read the CDF file - check the dimensions of the read in code within LISF/lis/dataassim/obs/NASA_AMSREsm, and see if it matches the reading of the CDF file in other DA plugins.

We have disabled the NASA AMSR-E plugin, and don't plan on any further support for it.  The AMSE-E data ends in 2011, and we are using more modern datasets.  Even the LPRM AMSR-E data is preferred over the NASA AMSR-E data, we have found.

**Reply by nchawang** [reply]
Hi @dmocko 
I see, okay.
Thankyou for the help and support!

**Comment by dmocko** [comment]
Hi @sujayvkumar and @yhkwon81 

Can you please look into this?  I am not familiar with this code for NASA_AMSRE soil moisture CDF generation.

One area to look may be inconsistencies between ldt/DAobs/NASA_AMSRE_sm/NASA_AMSREsm_obsMod.F90 and lis/dataassim/obs/NASA_AMSREsm/NASA_AMSREsm_Mod.F90

David
